src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleEqualityTests.cs

```diff
@@ -2248,6 +2286,148 @@ public static (A, A) GetTuple()
             CompileAndVerify(comp, expectedOutput: "A:1, GetTuple, A:30, A:40, ValueTuple2, A:4, X:5, X:6, Y:7, Y:8, X -> Y:5, A(1) == Y(5), X -> Y:6, A(30) == Y(6), A(40) == Y(7), A(4) == Y(8), True");
         }
 
+        [Fact]
+        public void TestConstrainedValueTuple()
+        {
+            var source = @"
+class C
+{
+    void M()
+    {
+        _ = (this, this) == (0, 1); // constraint violated by tuple in source
+        _ = (this, this) == (this, this); // constraint violated by converted tuple
+    }
+    public static bool operator ==(C c, int i)
+        => throw null;
+    public static bool operator !=(C c, int i)
+        => throw null;
+    public override bool Equals(object o)
+        => throw null;
+    public override int GetHashCode()
+        => throw null;
+    public static implicit operator int(C c)
+        => throw null;
+}
+namespace System
+{
+    public struct ValueTuple<T1, T2>
+        where T1 : class
+        where T2 : class
+    {
+        public T1 Item1;
+        public T2 Item2;
+
+        public ValueTuple(T1 item1, T2 item2)
+        {
+            this.Item1 = item1;
+            this.Item2 = item2;
+            throw null;
+        }
+    }
+}
+";
+
+            var comp = CreateCompilation(source);
+            comp.VerifyDiagnostics(
+                // (6,30): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T1' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "0").WithArguments("System.ValueTuple<T1, T2>", "T1", "int").WithLocation(6, 30),
+                // (6,33): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "1").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(6, 33),
+                // (6,30): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T1' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "0").WithArguments("System.ValueTuple<T1, T2>", "T1", "int").WithLocation(6, 30),
+                // (6,33): error CS0452: The type 'int' must be a reference type in order to use it as parameter 'T2' in the generic type or method 'ValueTuple<T1, T2>'
+                //         _ = (this, this) == (0, 1); // constraint violated by tuple in source
+                Diagnostic(ErrorCode.ERR_RefConstraintNotSatisfied, "1").WithArguments("System.ValueTuple<T1, T2>", "T2", "int").WithLocation(6, 33),
```

Comment:
What does the second pair of errors represent?

Reply:
There are tuple literals in source, and they have types (because both of the elements have types), except that they violate the constraint.
Then there are tuple conversions, which also have types and also violate constraints.

---
In reply to: [173845849](https://github.com/dotnet/roslyn/pull/25374#discussion_r173845849) [](ancestors = 173845849)